### ISSUE TITLE
Use DataflowResultsCursor instead of FlowAtLocation

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#62315]
#61922 introduced `DataflowResultsCursor`, which provides a stateful caching interface atop `FlowAtLocation`.

It would be better to use `DataflowResultsCursor` from more places, both because it's a more convenient / less error-prone interface, and because it's best to have more than one code path exercising the same code. It's also awkward trying to support both interfaces at once (the cursor implementation would be cleaner if we didn't have to).

The main drawback to using it is that there will be extra state/branching in places where we don't need it.

cc @eddyb who asked me to open this, and (along with @nikomatsakis) helped provide inspiration for `DataflowResultsCursor`
cc @ecstatic-morse @pnkfelix

**Comment by ecstatic-morse** [comment]
Here's a brief summary of the current state of things.

There's now 3 (and a half) places in `librustc_mir` which implement the same functionality: [`state_for_location`](https://github.com/rust-lang/rust/blob/848e0a23f34aaab3e4a974b031c86ef2a4e4fcc1/src/librustc_mir/dataflow/mod.rs#L456), [`DataflowResultsConsumer`](https://github.com/rust-lang/rust/blob/848e0a23f34aaab3e4a974b031c86ef2a4e4fcc1/src/librustc_mir/dataflow/mod.rs#L335)/[`FlowAtLocation`](https://github.com/rust-lang/rust/blob/848e0a23f34aaab3e4a974b031c86ef2a4e4fcc1/src/librustc_mir/dataflow/at_location.rs#L140) and the newly added `DataflowResultsCursor`. There's also [`rustc_peek`](https://github.com/rust-lang/rust/blob/848e0a23f34aaab3e4a974b031c86ef2a4e4fcc1/src/librustc_mir/transform/rustc_peek.rs#L195) (the half), which is similar but only considers the transfer function at assignment statements. However, this is just an (unnecessary) optimization; it could easily be rewritten to use a cursor instead.

`state_for_location` can also be replaced easily. `DataflowResultsConsumer`, however, is a bit more complex because it allows results to be zipped together (this is the purpose of [`FlowsAtLocation`](https://github.com/rust-lang/rust/blob/848e0a23f34aaab3e4a974b031c86ef2a4e4fcc1/src/librustc_mir/dataflow/at_location.rs#L17), although it claims to be for "cartesian products"). We could either rewrite all the consumers of this API to use the cursor API, or make `DataflowResultsCursor` the underlying implementation of `DataflowResultsConsumer` like so:

```rust
trait DataflowResultsVisitorEngine<'a> {
    type State: 'a;
    fn reset_to_start_of_block(&'a mut self, block: BasicBlock);
    fn seek(&'a mut self, loc: Location);
    fn get(&'a self) -> Self::State;
}

// This impl would be extended to the various tuple sizes
impl<'a, A, B> DataflowResultsVisitorEngine<'a> for (DataflowResultsCursor<A>, DataflowResultsCursor<B>) {
    type State = (&'a BitSet<A::Idx>, &'a BitSet<B::Idx>);
    /* ... */
}

trait DataflowResultsVisitor<'a, 'tcx> {
      type Engine: DataflowResultsVisitorEngine<'a>;

    // Observation Hooks: override (at least one of) these to get analysis feedback.
    fn visit_block_entry(&mut self,
                         _bb: BasicBlock,
                         _flow_state: Self::Engine::State) {}

    fn visit_statement_entry(&mut self,
                             _loc: Location,
                             _stmt: &Statement<'tcx>,
                             _flow_state: Self::Engine::State) {}

    fn visit_terminator_entry(&mut self,
                              _loc: Location,
                              _term: &Terminator<'tcx>,
                              _flow_state: Self::Engine::State) {}

    fn visit(&mut self, cursor: &'a mut Self::Engine) {
         todo!()
    }
}
```

`DataflowResultsVisitorEngine` is basically a streaming iterator with a few extra methods. You could almost do this with a generic `StreamingIterator` trait with a `zip` method, but that means `StreamingIterator::next` would have to visit statements, terminators, and the start of a block.

**Comment by jonas-schievink** [comment]
Seems like this is done, right?

**Comment by ecstatic-morse** [comment]
Resolved by #69644.
